### PR TITLE
C# 12: Primary constructors.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce library and data flow support for *primary* constructors.
Below we see an example of class `C1` with a primary constructor defined on line 1.
```csharp
public class C1(object o, string s)
{
    public object Obj => o;
    public string Str => s;

    public C1(object o) : this(o, "default") { }

    public void SetStr(string s1)
    {
        s = s1;
    }
}
```

The QL library now contains a class called `PrimaryConstructor` which contains exactly the primary constructors defined in source code. If it turns our that this implementation is insufficient we can extend the extractor to specifically extract whether a constructor is a primary constructor by inspecting the declaring syntax references using the the same "hack" as [roslyn](https://github.com/dotnet/roslyn/blob/580129a61f2cb88be53b7b225c00b8bd9af1d497/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ITypeSymbolExtensions.cs#L40).

Before discussing the details of the data flow implementation for primary constructors is is worth noting that the example class declaration of `C1` is roughly equivalent to (this is also what you will approximately get if you compile and de-compile the code):

```csharp
public class C1
{
    private object vObj;
    public object Obj => vObj;

    private string vStr;
    public string Str => vStr;

    public C1(object o, string s)
    {
        vObj = o;
        vStr = s;
    }

    public C1(object o) : this(o, "default") { }

    public void SetStr(string s1)
    {
        vStr = s1;
    }
}
```
That is, we can think of class with a primary constructor as having a backing field for each primary constructor parameter and that the primary constructor as having a (synthetic) body that assigns the provided arguments to the backing fields. 
Furthermore, each primary constructor parameter access (read/write) corresponds to reading or writing the corresponding backing field.
We use these observations to model data flow for primary constructors by introducing the following new content and dataflow node types

Content
- `PrimaryConstructorParameterContent`: As we don't have a "real" backing field for each primary constructor parameter we introduce a new content type that enables us to create "synthetic" backing fields for each primary constructor parameter.

The data-flow implementation introduces two new types of data-flow nodes:
- `PrimaryConstructorThisAccessNode`: Used for *this* access and synthesises the body of the primary constructor (that is, dataflow from the primary constructor parameters into their synthetic backing fields).
- `InstanceParameterAccessNode`: Used to synthesise reads and writes of the parameter synthetic backing fields. That is, for all primary constructor parameter accesses (reads and writes) within a class declaration we need to do the corresponding update to the synthetic backing fields.

The dataflow implementation has some limitations and doesn't cover e.g.
- Subtyping.

Huge thanks to @hvitved for outlining and helping with the implementation.

### CODE REVIEW

**Comment by michaelnebel** [comment]
DCA was uneventful; Even though it is a bit discouraging that we didn't find more results, doing improvements like this should help us avoid our dataflow analysis to deteriorate over time when the new language feature becomes more widely adopted.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -1881,6 +2000,16 @@ private PropertyContent getResultContent() {
...
 
+private predicate primaryConstructorParameterStore(Node node1, ContentSet c, Node node2) {
+  exists(AssignExpr ae, ParameterWrite pa, PrimaryConstructor constructor |
+    ae.getLValue() = pa and
+    pa.getTarget() = constructor.getAParameter() and
+    node1.asExpr() = ae.getRValue() and
+    node2 = TInstanceParameterAccessNode(ae.getAControlFlowNode(), true) and
+    c.(PrimaryConstructorParameterContent).getParameter() = pa.getTarget()
+  )
+}
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:2011]: Requires making `LocalExprStepConfiguration` public:

```suggestion
private predicate primaryConstructorParameterStore(
  Node node1, PrimaryConstructorParameterContent c, Node node2
) {
  exists(ControlFlow::Node cfn, Parameter p |
    node2 = TInstanceParameterAccessNode(cfn, true) and
    c.getParameter() = p
  |
    // direct assignment
    exists(LocalFlow::LocalExprStepConfiguration conf, AssignableDefinition def |
      conf.hasDefPath(_, node1.(ExprNode).getControlFlowNode(), def, cfn) and
      p = def.getTarget()
    )
    or
    // indirect assignment (for example as an `out` argument)
    exists(Ssa::ExplicitDefinition def |
      def = node1.(SsaDefinitionExtNode).getDefinitionExt() and
      p = def.getSourceVariable().getAssignable() and
      cfn = def.getControlFlowNode()
    )
  )
}
```

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -2010,6 +2147,12 @@ predicate readStep(Node node1, ContentSet c, Node node2) {
     node2.asExpr().(AwaitExpr).getExpr() = node1.asExpr() and
     c = getResultContent()
     or
+    exists(InstanceParameterAccessNode n | n = node1 |
+      n.getUnderlyingControlFlowNode() = node2.(ExprNode).getControlFlowNode() and
+      n.getParameter() = c.(PrimaryConstructorParameterContent).getParameter()
+    ) and
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:2153]:
```suggestion
    node1 = any(InstanceParameterAccessNode n |
      n.getUnderlyingControlFlowNode() = node2.(ExprNode).getControlFlowNode() and
      n.getParameter() = c.(PrimaryConstructorParameterContent).getParameter() and
      n.isPreUpdate()
    ) and
```

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -2072,6 +2215,10 @@ predicate clearsContent(Node n, ContentSet c) {
     c.(FieldContent).getField() = f.getUnboundDeclaration() and
     not f.isRef()
   )
+  or
+  exists(Node n1 |
+    primaryConstructorParameterStore(_, c, n1) and n = n1.(PostUpdateNode).getPreUpdateNode()
+  )
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:2221]:
```suggestion
  n = any(PostUpdateNode n1 | primaryConstructorParameterStore(_, c, n1)).getPreUpdateNode()
```

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 4]
```diff
@@ -2361,6 +2508,32 @@ module PostUpdateNodes {
 
     override Node getPreUpdateNode() { result.(FlowSummaryNode).getSummaryNode() = preUpdateNode }
   }
+
+  private class InstanceParameterAccessPostUpdateNode extends PostUpdateNode,
+    InstanceParameterAccessNode
+  {
+    private ControlFlow::Node cfg;
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:2515]: cfn

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll [hunk 5]
```diff
@@ -239,6 +239,23 @@ class PropertyContent extends Content, TPropertyContent {
...
+ */
+class PrimaryConstructorParameterContent extends Content, TPrimaryConstructorParameterContent {
+  private Parameter p;
+
+  PrimaryConstructorParameterContent() { this = TPrimaryConstructorParameterContent(p) }
+
+  /** Gets the underlying parameter. */
+  Parameter getParameter() { result = p }
+
+  override string toString() { result = "parameter field " + p.getName() }
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll:254]: I would just say `result = "parameter " + p.getName()`

File: csharp/ql/test/library-tests/constructors/PrimaryConstructor.ql [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/**
+ * @name Test for primary constructors
+ */
```
Comment (hvitved) [inline csharp/ql/test/library-tests/constructors/PrimaryConstructor.ql:3]: No need to add this.

**Review by hvitved — CHANGES_REQUESTED** [review]
Overall LGTM, great work! Some (mostly cosmetic) suggestions.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 7]
```diff
@@ -1795,6 +1794,19 @@ class FlowSummaryNode extends NodeImpl, TFlowSummaryNode {
...
+ * ```csharp
+ * public class C(object o)
+ * {
+ *    public object GetParam() => o; // (1)
+ *
+ *    public void SetParam(object value) => o = value; // (2)
+ * }
+ * ```
+ * the first access to o (1) is modeled as this.o_backing_field and
+ * the second access to o (2) is modeled as this.o_backing_field = value.
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1807]: Missing backticks.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 8]
```diff
@@ -1795,6 +1794,19 @@ class FlowSummaryNode extends NodeImpl, TFlowSummaryNode {
...
+ * Both models need a pre-update this node, and the latter need an additional post-update this access,
+ * all of which are represented by an InstanceParameterAccessNode node.
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1809]: Missing backticks.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 9]
```diff
@@ -1836,6 +1848,16 @@ class InstanceParameterAccessNode extends NodeImpl, TInstanceParameterAccessNode
 
 /**
  * A data-flow node used to synthesize the body of a primary constructor.
+ *
+ * For example, in
+ * ```csharp
+ * public class C(object o) { }
+ * ```
+ * we synthesize the primary constructor for C as
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1856]: Missing backticks.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 10]
```diff
@@ -1836,6 +1848,16 @@ class InstanceParameterAccessNode extends NodeImpl, TInstanceParameterAccessNode
...
+ * ```csharp
+ * public C(object o) => this.o_backing_field = o;
+ * ```
+ * The synthesized (pre/post-update) this access is represented an PrimaryConstructorThisAccessNode node.
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1860]: Missing backticks.

**Review by michaelnebel — COMMENTED** [review]
@hvitved : Your changes look plausible to me.

**Review by hvitved — APPROVED** [review]

**Pull request merged**